<a href="https://colab.research.google.com/github/saivarunvanama/data-using-python/blob/main/week2_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task
Analyze student test performance by creating original score and bonus matrices, calculating final scores with bonuses and penalties, and transposing the final scores matrix.

## Create original score matrix

### Subtask:
Generate a 3x3 matrix representing the original scores of students in different subjects.


**Reasoning**:
The first step is to import the numpy library and create the 3x3 matrix as instructed.



In [2]:
import numpy as np

original_scores = np.array([[85, 90, 78],
                            [92, 88, 95],
                            [76, 85, 80]])
print(original_scores)

[[85 90 78]
 [92 88 95]
 [76 85 80]]


## Design bonus marks matrix

### Subtask:
Generate a 3x3 matrix representing the bonus marks awarded to students in different subjects.


**Reasoning**:
Create a 3x3 NumPy array for bonus marks and print it.



In [3]:
bonus_marks = np.array([[5, 3, 7],
                       [6, 4, 8],
                       [2, 5, 3]])
print(bonus_marks)

[[5 3 7]
 [6 4 8]
 [2 5 3]]


## Calculate final scores with bonuses

### Subtask:
Add the bonus marks matrix to the original score matrix to get the scores after adding bonuses.


**Reasoning**:
Add the `original_scores` and `bonus_marks` matrices and store the result in `scores_with_bonuses`. Then print the result to verify.



In [4]:
scores_with_bonuses = original_scores + bonus_marks
print(scores_with_bonuses)

[[ 90  93  85]
 [ 98  92 103]
 [ 78  90  83]]


## Apply penalty for late submissions

### Subtask:
Multiply the scores (with bonuses) by a penalty factor to simulate a penalty for late submissions.


**Reasoning**:
Define the penalty factor, multiply the scores with bonuses by the penalty factor, store the result, and print the final scores.



In [5]:
penalty_factor = 0.9
final_scores = scores_with_bonuses * penalty_factor
print(final_scores)

[[81.  83.7 76.5]
 [88.2 82.8 92.7]
 [70.2 81.  74.7]]


## Transpose the final scores matrix

### Subtask:
Transpose the final scores matrix to easily analyze performance by subject.


**Reasoning**:
Transpose the final scores matrix and print the result.



In [6]:
transposed_final_scores = final_scores.T
print(transposed_final_scores)

[[81.  88.2 70.2]
 [83.7 82.8 81. ]
 [76.5 92.7 74.7]]


## Summary:

### Data Analysis Key Findings

*   The original scores matrix was created as a 3x3 array: \begin{bmatrix} 85 & 90 & 78 \\ 92 & 88 & 95 \\ 76 & 85 & 80 \end{bmatrix}.
*   The bonus marks matrix was created as a 3x3 array: \begin{bmatrix} 5 & 3 & 7 \\ 6 & 4 & 8 \\ 2 & 5 & 3 \end{bmatrix}.
*   The scores after adding bonuses were calculated by adding the original scores and bonus marks matrices element-wise: \begin{bmatrix} 90 & 93 & 85 \\ 98 & 92 & 103 \\ 78 & 90 & 83 \end{bmatrix}.
*   A penalty factor of 0.9 was applied to the scores with bonuses, resulting in the final scores matrix: \begin{bmatrix} 81.0 & 83.7 & 76.5 \\ 88.2 & 82.8 & 92.7 \\ 70.2 & 81.0 & 74.7 \end{bmatrix}.
*   The final scores matrix was transposed to facilitate analysis by subject: \begin{bmatrix} 81.0 & 88.2 & 70.2 \\ 83.7 & 82.8 & 81.0 \\ 76.5 & 92.7 & 74.7 \end{bmatrix}.

### Insights or Next Steps

*   The transposed matrix allows for a quick comparison of how different students performed in each specific subject after all adjustments.
*   Further analysis could involve calculating average scores per subject or per student from the transposed matrix to identify areas of strength or weakness.
